src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt

```diff
@@ -1,124 +1,225 @@
+        .before(WHERE_CLAUSE).spacing(1, 1, 0, true, 0)
+        .applyForEach(BLOCK_LIKE) { before(it).spaces(1) }
+        .beforeInside(LBRACE, ts(FOREIGN_MOD_ITEM, MOD_ITEM)).spaces(1)
+
+        .between(ts(IDENTIFIER, FN), PARAMS_LIKE).spaceIf(false)
+        .between(IDENTIFIER, GENERIC_PARAMS).spaceIf(false)
+        .between(IDENTIFIER, GENERIC_ARGS).spaceIf(false)
+        .between(IDENTIFIER, ARG_LIST).spaceIf(false)
+        .between(GENERIC_PARAMS, PARAMS_LIKE).spaceIf(false)
+        .beforeInside(ARG_LIST, CALL_EXPR).spaceIf(false)
+
+        .between(BINDING_MODE, IDENTIFIER).spaces(1)
+        .between(IMPL, GENERIC_PARAMS).spaces(0)
+        .afterInside(GENERIC_PARAMS, IMPL_ITEM).spaces(1)
+        .betweenInside(ts(GENERIC_PARAMS), TYPES, IMPL_ITEM).spaces(1)
+
+        .afterInside(LBRACE, BLOCK_LIKE).spacing(1, 1, 0, true, 0)
+        .beforeInside(RBRACE, BLOCK_LIKE).spacing(1, 1, 0, true, 0)
+
+        .betweenInside(IDENTIFIER, ALIAS, EXTERN_CRATE_ITEM).spaces(1)
+
+        .betweenInside(IDENTIFIER, ENUM_TUPLE_ARGS, ENUM_VARIANT).spaces(0)
+        .betweenInside(IDENTIFIER, ENUM_DISCRIMINANT, ENUM_VARIANT).spaces(1)
+
+        //== types
+        .afterInside(LIFETIME, REF_TYPE).spaceIf(true)
+        .betweenInside(ts(MUL), ts(CONST, MUT), PTR_TYPE).spaces(0)
+        .before(TYPE_PARAM_BOUNDS).spaces(0)
+        .beforeInside(LPAREN, PATH_PART).spaces(0)
+
+        //== expressions
+        .beforeInside(LPAREN, PAT_ENUM).spaces(0)
+        .beforeInside(LBRACK, INDEX_EXPR).spaces(0)
+        .afterInside(PARAMS_LIKE, LAMBDA_EXPR).spacing(1, 1, 0, true, 1)
+
+        //== macros
+        .betweenInside(IDENTIFIER, EXCL, MACRO_INVOCATION).spaces(0)
+        .between(MACRO_INVOCATION, MACRO_ARGS).spaces(0)
+        .betweenInside(MACRO_INVOCATION, IDENTIFIER, MACRO_DEFINITION).spaces(1)
+        .betweenInside(IDENTIFIER, MACRO_ARG, MACRO_DEFINITION).spaces(1)
+
+        //== rules with very large area of application
+        .around(NO_SPACE_AROUND_OPS).spaces(0)
+        .around(SPACE_AROUND_OPS).spaces(1)
+        .around(KEYWORDS).spaces(1)
+
+fun computeSpacing(parentBlock: Block, child1: Block?, child2: Block, ctx: RustFmtBlockContext): Spacing? {
+    if (child1 is AbstractRustFmtBlock && child2 is AbstractRustFmtBlock) {
+        val origNode1 = child1.node;
+        val origNode2 = child2.node;
+        val origPsi1 = origNode1.psi
+        val origPsi2 = origNode2.psi
+        val parentPsi = origPsi1.parent
+        val parentParentPsi = parentPsi.parent
+
+        // Here are rules which are comment-sensitive
+        when {
+            origPsi1 is RustOuterAttr && origPsi2 is PsiComment
+            -> return one()
+
+            origPsi1 is RustOuterAttr && (origPsi2 is RustOuterAttr || origPsi1.parent is RustItem)
+                || origPsi1 is PsiComment && (origPsi2 is RustOuterAttr || origPsi1.getPrevNonCommentSibling() is RustOuterAttr)
```

Comment:
The hardest thing in programming is naming things. The second hardest it beautifully laying out complex logical expressions :|